messaging/broker.go

```diff
@@ -144,943 +222,944 @@ func (b *Broker) closeTopics() {
+func (fsm *RaftFSM) Snapshot(w io.Writer) (uint64, error) { return fsm.Broker.Snapshot(w) }
+func (fsm *RaftFSM) Restore(r io.Reader) error            { return fsm.Broker.Restore(r) }
 
-	// Close replica's writer.
-	r.closeWriter()
+// MustApply applies a raft command to the broker. Panic on error.
+func (fsm *RaftFSM) MustApply(e *raft.LogEntry) {
+	switch e.Type {
+	case raft.LogEntryCommand:
+		// Decode message.
+		m := &Message{}
+		if err := m.UnmarshalBinary(e.Data); err != nil {
+			panic("message unmarshal: " + err.Error())
+		}
+		m.Index = e.Index
 
-	// Remove replica from broker.
-	delete(b.replicas, c.ID)
+		// Apply message.
+		if err := fsm.Broker.Apply(m); err != nil {
+			panic(err.Error())
+		}
 
-	b.mustSave()
+	default:
+		// Move internal index forward if it's an internal raft comand.
+		if err := fsm.Broker.SetMaxIndex(e.Index); err != nil {
+			panic(fmt.Sprintf("set max index: idx=%d, err=%s", e.Index, err))
+		}
+	}
 }
 
-// Subscribe adds a subscription to a topic from a replica.
-func (b *Broker) Subscribe(replicaID, topicID uint64) error {
-	b.mu.Lock()
-	defer b.mu.Unlock()
+// DefaultMaxSegmentSize is the largest a segment can get before starting a new segment.
+const DefaultMaxSegmentSize = 10 * 1024 * 1024 // 10MB
 
-	// TODO: Allow non-zero starting index.
+// topic represents a single named queue of messages.
+// Each topic is identified by a unique path.
+//
+// Topics write their entries to segmented log files which contain a
+// contiguous range of entries.
+type Topic struct {
+	mu    sync.Mutex
+	id    uint64 // unique identifier
+	index uint64 // highest index replicated
+	path  string // on-disk path
 
-	// Ensure replica & topic exist.
-	if b.replicas[replicaID] == nil {
-		return ErrReplicaNotFound
-	}
+	file   *os.File // last segment writer
+	opened bool
 
-	// Issue command to subscribe to topic.
-	return b.PublishSync(&Message{
-		Type: SubscribeMessageType,
-		Data: mustMarshalJSON(&SubscribeCommand{ReplicaID: replicaID, TopicID: topicID}),
-	})
+	// The largest a segment can get before splitting into a new segment.
+	MaxSegmentSize int64
 }
 
-func (b *Broker) mustApplySubscribe(m *Message) {
-	var c SubscribeCommand
-	mustUnmarshalJSON(m.Data, &c)
+// NewTopic returns a new instance of Topic.
+func NewTopic(id uint64, path string) *Topic {
+	return &Topic{
+		id:   id,
+		path: path,
 
-	// Retrieve replica.
-	r := b.replicas[c.ReplicaID]
-	if r == nil {
-		return
-	}
-
-	// Save current index on topic.
-	t := b.createTopicIfNotExists(c.TopicID)
-
-	// Ensure topic is not already subscribed to.
-	if _, ok := r.topics[c.TopicID]; ok {
-		b.Logger.Printf("already subscribed to topic: replica=%d, topic=%d", r.id, c.TopicID)
-		return
+		MaxSegmentSize: DefaultMaxSegmentSize,
 	}
+}
 
-	// Add subscription to replica.
-	r.topics[c.TopicID] = c.Index
-	t.replicas[c.ReplicaID] = r
+// ID returns the topic identifier.
+func (t *Topic) ID() uint64 { return t.id }
 
```

Comment:
Yes, if the field gets set in a constructor then it should be unexported and accessible through an accessor IMO.